drivers/sensor/bosch/bmm350/bmm350.c

```diff
@@ -0,0 +1,984 @@
+		LOG_ERR("failed to read sample");
+		return -EIO;
+	}
+	drv_data->mag_temp_data.x = mag_temp_data.x;
+	drv_data->mag_temp_data.y = mag_temp_data.y;
+	drv_data->mag_temp_data.z = mag_temp_data.z;
+	return 0;
+}
+
+/*
+ * ut change to Gauss
+ */
+static void bmm350_convert(struct sensor_value *val, int raw_val)
+{
+	val->val1 = raw_val / 100;
+	val->val2 = (raw_val % 100) * 10000;
+}
+
+static int bmm350_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct bmm350_data *drv_data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_MAGN_X:
+		bmm350_convert(val, drv_data->mag_temp_data.x);
+		break;
+	case SENSOR_CHAN_MAGN_Y:
+		bmm350_convert(val, drv_data->mag_temp_data.y);
+		break;
+	case SENSOR_CHAN_MAGN_Z:
+		bmm350_convert(val, drv_data->mag_temp_data.z);
+		break;
+	case SENSOR_CHAN_MAGN_XYZ:
+		bmm350_convert(val, drv_data->mag_temp_data.x);
+		bmm350_convert(val + 1, drv_data->mag_temp_data.y);
+		bmm350_convert(val + 2, drv_data->mag_temp_data.z);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+static uint8_t acc_odr_to_reg(const struct sensor_value *val)
+{
+	double odr = sensor_value_to_double((struct sensor_value *)val);
+
+	uint8_t reg = BMM350_DATA_RATE_100HZ;
+
+	if ((odr >= 0.78125) && (odr <= 1.5625)) {
+		reg = BMM350_DATA_RATE_1_5625HZ;
+	} else if ((odr > 1.5625) && (odr <= 3.125)) {
+		reg = BMM350_DATA_RATE_3_125HZ;
+	} else if ((odr > 3.125) && (odr <= 6.25)) {
+		reg = BMM350_DATA_RATE_6_25HZ;
+	} else if ((odr > 6.25) && (odr <= 12.5)) {
+		reg = BMM350_DATA_RATE_12_5HZ;
+	} else if ((odr > 12.5) && (odr <= 25.0)) {
+		reg = BMM350_DATA_RATE_25HZ;
+	} else if ((odr > 25.0) && (odr <= 50.0)) {
+		reg = BMM350_DATA_RATE_50HZ;
+	} else if ((odr > 50.0) && (odr <= 100.0)) {
+		reg = BMM350_DATA_RATE_100HZ;
+	} else if ((odr > 100.0) && (odr <= 200.0)) {
+		reg = BMM350_DATA_RATE_200HZ;
+	} else if ((odr > 200.0) && (odr <= 400.0)) {
+		reg = BMM350_DATA_RATE_400HZ;
+	} else if (odr > 400.0) {
+		reg = BMM350_DATA_RATE_400HZ;
+	}
+	return reg;
+}
+
+static uint8_t acc_osr_to_reg(const struct sensor_value *val)
+{
+	switch (val->val1) {
+	case 0:
+		return BMM350_NO_AVERAGING;
+	case 2:
+		return BMM350_AVERAGING_2;
+	case 4:
+		return BMM350_AVERAGING_4;
+	case 8:
+		return BMM350_AVERAGING_8;
+	default:
+		return 0xFF;
+	}
+}
+
+/*!
+ * @brief This API sets the ODR and averaging factor.
+ */
+int8_t bmm350_set_odr_performance(enum bmm350_data_rates odr,
+				  enum bmm350_performance_parameters performance,
+				  const struct device *dev)
+{
+	/* Variable to store the function result */
+	int8_t rslt = 0x00;
+	/* Variable to get PMU command */
+	uint8_t reg_data = 0;
+	enum bmm350_performance_parameters performance_fix = performance;
+
+	if (rslt == BMM350_OK) {
+		/* Reduce the performance setting when too high for the chosen ODR */
+		if ((odr == BMM350_DATA_RATE_400HZ) && (performance >= BMM350_AVERAGING_2)) {
```

Comment:
This sort of behind the back side effect when changing one attribute and then having to go modify others is exactly why I wrote the icm42688 the way I did with an all or nothing config and rollback.

Either the desired configuration is valid or not, and if not then setting an attribute should fail and the sensor go back to the way it was configured before. Attributes should likely be understood to *not* have side effects, but that they *may* require ordering in my very humble opinion.

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/tdk/icm42688/icm42688.c#L123

Reply:
what would you like me to change here? only certain combinations of ODR and OSR are allowed.

Reply:
Argh, that is painful. I don't think we have a great answer to this sort of inter-dependency with attributes, but its a recurring thing for sure. No worries, I see the warning message there I guess that's as good as we can do today.